### PR TITLE
test spaces

### PR DESCRIPTION
**Pull request by DvirDukhan**
test spaces for:
1. sanity check
2. increase coverage

### CODE REVIEW

File: tests/unit/test_spaces.cpp [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    case ARCH_OPT_AVX:
+        ASSERT_EQ(baseline, InnerProductSIMD4ExtResiduals_AVX(v, v, &dim));
+        optimization = ARCH_OPT_SSE;
+    case ARCH_OPT_SSE:
+        ASSERT_EQ(baseline, InnerProductSIMD4ExtResiduals_SSE(v, v, &dim));
+        break;
+    default:
+        ASSERT_TRUE(false);
+    }
+}
```
Comment (alonre24) [inline tests/unit/test_spaces.cpp:122]: new line

File: tests/unit/test_spaces.cpp [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        optimization = ARCH_OPT_SSE;
+    case ARCH_OPT_SSE:
+        ASSERT_EQ(baseline, L2SqrSIMD16ExtResiduals_SSE(v, v, &dim));
+        break;
+    default:
+        ASSERT_TRUE(false);
+    }
+}
+
+// This test will trigger the "Residuals" function for dimension < 9, for each optimization.
```
Comment (alonre24) [inline tests/unit/test_spaces.cpp:49]: Why "for dimension < 9"? The dim in this test is 9...

File: tests/unit/test_spaces.cpp [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    ~SpacesTest() override {}
+
+    void SetUp() override {}
+
+    void TearDown() override {}
+};
+
+// This test will trigger the "Residuals" function for dimension > 16, for each optimization.
+TEST_F(SpacesTest, l2_17) {
+    Arch_Optimization optimization = getArchitectureOptimization();
```
Comment (alonre24) [inline tests/unit/test_spaces.cpp:26]: We guarantee that `optimization` is `ARCH_OPT_AVX512` in CI?

File: tests/unit/test_spaces.cpp [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    switch (optimization) {
+    case ARCH_OPT_AVX512:
+        ASSERT_EQ(baseline, L2SqrSIMD16ExtResiduals_AVX512(v, v, &dim));
+        optimization = ARCH_OPT_AVX;
+    case ARCH_OPT_AVX:
+        ASSERT_EQ(baseline, L2SqrSIMD16ExtResiduals_AVX(v, v, &dim));
+        optimization = ARCH_OPT_SSE;
+    case ARCH_OPT_SSE:
+        ASSERT_EQ(baseline, L2SqrSIMD4ExtResiduals_SSE(v, v, &dim));
+        break;
```
Comment (alonre24) [inline tests/unit/test_spaces.cpp:68]: What is the difference from the previous test?

**Review by alonre24 — COMMENTED** [review]
Nice, few questions. Also, does it cover all of our distance function variations? I think that there are some that aren't covered, no?

File: tests/unit/test_spaces.cpp (hunk 3 above)
Comment (DvirDukhan) [reply tests/unit/test_spaces.cpp:26]: not for now. I'll explain f2f

File: tests/unit/test_spaces.cpp (hunk 4 above)
Comment (DvirDukhan) [reply tests/unit/test_spaces.cpp:68]: the vector length compared to 16 and 4. This is how to space choses the best function

**Review by alonre24 — APPROVED** [review]
Cool

**Pull request merged**